src/controllers/admin.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{
+    app::AppState,
+    auth::AuthCheck,
+    models::{CrateOwner, OwnerKind, User, Version},
+    schema::*,
+    util::errors::{AppResult, custom},
+};
+use axum::{Json, extract::Path};
+use chrono::{DateTime, Utc};
+use diesel::{dsl::count_star, prelude::*};
+use diesel_async::RunQueryDsl;
+use http::{StatusCode, request::Parts};
+use serde::Serialize;
+use std::collections::HashMap;
+
+/// Handles the `GET /api/private/admin_list/{username}` endpoint.
+pub async fn list(
+    state: AppState,
+    Path(username): Path<String>,
+    req: Parts,
+) -> AppResult<Json<AdminListResponse>> {
+    let mut conn = state.db_write().await?;
+
+    let auth = AuthCheck::default().check(&req, &mut conn).await?;
+    let logged_in_user = auth.user();
+
+    if !logged_in_user.is_admin {
+        return Err(custom(
+            StatusCode::FORBIDDEN,
+            "must be an admin to use this route",
+        ));
+    }
+
+    let (user, verified, email) = users::table
+        .left_join(emails::table)
+        .filter(users::gh_login.eq(username))
+        .select((
+            User::as_select(),
+            emails::verified.nullable(),
+            emails::email.nullable(),
+        ))
+        .first::<(User, Option<bool>, Option<String>)>(&mut conn)
+        .await?;
+
+    let crates: Vec<(i32, String, DateTime<Utc>, i64)> = CrateOwner::by_owner_kind(OwnerKind::User)
+        .inner_join(crates::table)
+        .filter(crate_owners::owner_id.eq(user.id))
+        .select((
+            crates::id,
+            crates::name,
+            crates::updated_at,
+            rev_deps_subquery(),
+        ))
+        .order(crates::name.asc())
+        .load(&mut conn)
+        .await?;
+
+    let crate_ids: Vec<_> = crates.iter().map(|(id, ..)| id).collect();
+
+    let versions: Vec<Version> = versions::table
+        .filter(versions::crate_id.eq_any(crate_ids))
+        .select(Version::as_select())
+        .load(&mut conn)
+        .await?;
+    let mut versions_by_crate_id: HashMap<i32, Vec<Version>> = HashMap::new();
+    for version in versions {
+        let crate_versions = versions_by_crate_id.entry(version.crate_id).or_default();
+        crate_versions.push(version);
+    }
```

Comment:
if we use the default version instead of the last version then we only need the versions list for the "number of versions" field and that information is also available in the `default_versions` table. in other words it could be folded in the crates query above.